blog/content/posts/2025-12-xx-transport-direction.md

```diff
@@ -0,0 +1,83 @@
+
+This means that MCP applications will be able to handle sessions as part of their domain logic. A cookie style mechanism is the preferred choice.
+
+This makes MCP similar to standard HTTP - where the protocol itself is stateless, while applications build stateful semantics with cookies, tokens and similar mechanisms. The exact approach to session creation is still being worked on - but this removes existing ambiguities on what a session is in remote MCP.
+
+### Elicitations and Sampling
+
+Two standout MCP features enable advanced AI workflows: [Elicitations](https://spec.modelcontextprotocol.io/specification/2025-03-26/client/elicitation) for requesting human input, and [Sampling](https://spec.modelcontextprotocol.io/specification/2025-03-26/client/sampling) for agentic LLM interactions.
+
+In our stateless design, this bidirectional request pattern requires rethinking. Currently when a Server needs more information to complete a Tool Call, it suspends operation and waits for a Client response - meaning it must remember all of its outstanding requests.
+
+To avoid the need to do this, we'll make the Server Request/Response similar to the way Chat APIs work. This means that the Client will return both the Request _and_ Response allowing the Server to reconstruct the necessary state purely from the returned message.
+
+### Update Notifications, Subscriptions
+
+MCP's dynamic nature means that Tools, Prompts, and Resources can change during operation. The current protocol uses server-to-client `ListChangedNotification` messages as an optimization to prompt cache invalidation.
+
+We're replacing the general-purpose GET stream with explicit subscription streams. Clients will start specialized streams when they want to subscribe to specific items, and can manage multiple concurrent subscriptions. If a subscription stream is interrupted, the client simply restarts it—no complex resumption logic required.
+
+To make notifications truly optional optimizations rather than requirements, we're adding TTLs and version numbers (potentially using etags) to data. This allows clients to make intelligent caching decisions without depending on the notification stream, improving reliability in lossy network conditions.
+
+### JSON-RPC Envelopes
+
+The protocol currently uses JSON-RPC for all message envelopes, including method names and parameters. As we optimize for HTTP deployments, questions arise about whether to move toward more traditional REST patterns.
+
+While we don't yet have consensus on fully replacing JSON-RPC with REST, we agree that copying the JSON-RPC method name into the HTTP path (or a dedicated header) improves clarity and enables better HTTP caching semantics.
+
+The core tension is between protocol consistency (keeping JSON-RPC everywhere) and HTTP integration (making MCP look like normal web traffic). We're evaluating this tradeoff with real deployment scenarios in mind.
```

Comment:
IIUC the discussion correctly, I think the argument is less about protocol consistency (between stdio and http) and more about consistency between new versions and olds ones (minimizing the work of still-evolving infra to keep up with MCP)